documentation.md

```diff
@@ -0,0 +1,669 @@
+
+### Finding optimal entries
+
+To find optimal the optimal entries to simulate new memory operations, the entries are first grouped by rank.
+All these groups, quotiented by rank, work in parallel without interaction.
+
+Entries are then regrouped by cost categories (which are representations of memory operation latency on fewer bits), depending on the memory operation address and on the current corresponding rank row buffer state.
+
+The oldest candidate entry for each cost category is determined by masking the main age matrix rows with the _matches_cond_ signal.
+
+The lowest cost category where there is at least one candidate entry is determined.
+The oldest candidate within this cost category
+
+Additionally, the cost and row buffer identifier corresponding to the optimal entry are determined (resp. on the signals _opti_cost_cat_ and _opti_rbuf_).
+This row buffer identifier helps updating the simulated rank state.
+
+### Detailed slot operation
+
+#### Request acceptance
+
+##### Write requests
+
+When a write slot is free (_~v_), a new write address request can be accepted.
+The allocated free slot fields takes the following values:
+
+- _v_ is set.
+- _iid_ takes the value held by the _waddr_iid_i_ signal, which is the write reservation output signal from the response banks.
+- _addr_ takes the value held by the address request.
+- _burst_size_ takes the value held by the address request.
+- _burst_fixed_ takes the value 1 iff the burst type is fixed.
+- _data_v_ takes the value, for each bit:
+  - 1 for the first _wdata_immediate_cnt_i_ bits, which correspond to the number of write data requests that had arrived not later than the write address request.
+  - 0 for the next bits corresponding to entries in the burst.
+  - 1 for the last entries, which correspond to entries beyond the burst length.
+- _mem_pending_ takes the value '0, as no simulated memory operation has started for any of these new entries.
+- _mem_done_ is set to zero, except for the last bits, which correspond to entries beyond the burst length.
+  The latter bits are set, because an unset bit in the _mem_done_ array represents an entry which must be eventually completed.
+
+<figure class="image">
+  <img src="https://i.imgur.com/eUkYbI9.png" alt="Write slot state">
+  <figcaption>Fig: Write slot state just after acceptance of a new write address of burst length 2 and 1 immediate wdata (write data arrived not after the write address request)</figcaption>
+</figure>
+
+##### Read requests
+
+Read request acceptance is identical to write address acceptance, except that:
+
+- Read and write slots are disjoint.
+- The _data_v_ field is absent from read slots.
+  Virtually, all the bits would be set to 1 when a read slot is allocated.
+
+#### Entry and slot liberation
+
+For each write or read slot, for each entry, if the corresponding _mem_pending_ bit is set, then it is swapped with the _mem_done_ bit when the corresponding rank decrementing counter reaches 3.
```

Comment:
"For every entry in every write and read slot, if the _mem_pending_ bit is set, it is swapped with the _mem_done_ bit when the corresponding rank decrementing counter reaches 3."

Reply:
It is not clear to me, what the effect of this is and why it is needed. It seems that you do this because there is a delay between this unit and the response bank and you want the specified delay to be exact. But why would you swap the _mem_pending_ with _mem_done_? Do you want to say that if _mem_pending_ is set and the counter reaches 3 (meaning we will release in 3 cycles from now), you already change _mem_pending_ to 0 and _mem_done_ to 1 (as far as I understood this effectively finishes the request in the delay module) to make sure the response bank sees this in 3 cycles from now?

Reply:
Done!